base/shell/explorer/taskswnd.cpp

```diff
@@ -468,12 +528,199 @@ class CTaskSwitchWnd :
+            bandSite->AddBand(shellMenu);
+
+            IUnknown_SetSite(menuPopup, pSms);
+
+            CComPtr<IInitializeObject> pIo;
+            hr = menuPopup->QueryInterface(IID_PPV_ARG(IInitializeObject, &pIo));
+            if (SUCCEEDED(hr))
+                pIo->Initialize();
+        } else {
+            TaskGroupOpened = TaskGroup->Index;
+
+            menuPopup->OnSelect(MPOS_FULLCANCEL);
+
+            shellMenu->SetMenu(hMenu, NULL, SMSET_BOTTOM);
+        }
+
+        CComPtr<IDeskBand> pIDB;
+        hr = shellMenu->QueryInterface(IID_PPV_ARG(IDeskBand, &pIDB));
+        if(!SUCCEEDED(hr))
+            return;
+
+        HWND hWnd;
+        if (!SUCCEEDED(pIDB->GetWindow(&hWnd)))
+            return;
+
+        ::PostMessageW(hWnd, TB_SETIMAGELIST, 0, (LPARAM)m_TaskGroupImageList);
+
+        POINTL p = {test.left + test2.left,test.top};
+        menuPopup->Popup(&p, &test, MPPF_TOP | MPPF_SETFOCUS);
+
+        m_ActiveTaskItem = NULL;
+
+        m_TaskBar.CheckButton(TaskGroup->Index, TRUE);
+    }
+
+    INT GetTaskGroupItemIconIndex(IN INT Index) {
+        if (shellMenu == NULL)
+            return -1;
+
+        PTASK_ITEM TaskItem, LastTaskItem = NULL;
+
+        TaskItem = m_TaskItems;
+        LastTaskItem = TaskItem + m_TaskItemCount;
+
+        while (TaskItem != LastTaskItem)
+        {
+            if (TaskItem->Group->Index == TaskGroupOpened && TaskItem->GroupIndex == Index)
+            {
+                if (TaskItem->GroupIconIndex < 0)
+                {
+                    HICON icon = GetWndIcon(TaskItem->hWnd);
+                    if (!icon)
+                        icon = static_cast<HICON>(LoadImageW(NULL, MAKEINTRESOURCEW(OIC_SAMPLE), IMAGE_ICON, 0, 0, LR_SHARED | LR_DEFAULTSIZE));
+
+                    TaskItem->GroupIconIndex = ImageList_ReplaceIcon(m_TaskGroupImageList, -1, icon);
+                }
+
+                return TaskItem->GroupIconIndex;
+            }
+
+            TaskItem++;
+        }
+
+        return -1;
+    }
 
     INT UpdateTaskGroupButton(IN PTASK_GROUP TaskGroup)
     {
         ASSERT(TaskGroup->Index >= 0);
 
-        /* FIXME: Implement */
+        TBBUTTONINFO tbbi = { 0 };
+        HICON icon;
+        WCHAR windowText[255];
+
+        /* Open process to retrieve the filename of the executable */
+        HANDLE hTaskProc = OpenProcess(PROCESS_QUERY_INFORMATION | PROCESS_VM_READ, FALSE, TaskGroup->dwProcessId);
+        if(hTaskProc != NULL) {
+            WCHAR ExePath[MAX_PATH] = {};
+
+            if(GetModuleFileNameExW(hTaskProc, NULL, ExePath, _countof(ExePath)) && GetVersionInfoString(ExePath, L"FileDescription", windowText, _countof(windowText))) {
+                tbbi.pszText = windowText;
+            }
+
+            if (ExtractIconExW(ExePath, 0, NULL, &icon, 1) <= 0)
+                icon = static_cast<HICON>(LoadImageW(NULL, MAKEINTRESOURCEW(OIC_SAMPLE), IMAGE_ICON, 0, 0, LR_SHARED | LR_DEFAULTSIZE));
+
+            CloseHandle(hTaskProc);
```

Comment:
You are getting the path of a process in multiple places, please create a helper function to do this. This will help NT6 migration that has `QueryFullProcessImageNameW` without the need for PSAPI.

Reply:
True, will work on that

Reply:
This should be resolved now